### PR TITLE
ROCm: Support RCCL

### LINKED ISSUE #3927: ROCm: Support RCCL
RCCL is the ROCm's counterpart for NCCL:
https://github.com/ROCmSoftwarePlatform/rccl

### PR DESCRIPTION
**Pull request by leofang**
Close #3927. All tests in `tests/cupy_tests/cuda_tests/test_nccl.py` passed locally, but I only have 1 GPU, so we might need further tests to ensure it's working with multi-GPUs.

This works because RCCL uses exactly the same API names as in NCCL, so we don't even need a shim layer to wrap the names!

### CODE REVIEW

**Comment by leofang** [comment]
cc: @emcastillo

**Comment by leofang** [comment]
Jenkins, test this please

File: install/build.py [hunk 1]
```diff
@@ -155,7 +155,10 @@ def get_compiler_setting(use_hip):
         extra_compile_args.append('-std=c++11')
 
     if PLATFORM_DARWIN:
-        library_dirs.append('/usr/local/cuda/lib')
+        if cuda_path:
+            library_dirs.append('/usr/local/cuda/lib')
+        elif rocm_path:
+            library_dirs.append('/opt/rocm/lib')
```
Comment (leofang) [inline install/build.py:161]: Oops, it's Darwin...I think these lines will be removed in #3857.

**Comment by leofang** [comment]
Fixed a bug in f174fc2 that I wasn't able to catch because I have a CUDA installation on the development machine.

**Comment by leofang** [comment]
Jenkins, test this please

File: install/build.py [hunk 2]
```diff
@@ -404,22 +407,29 @@ def check_nccl_version(compiler, settings):
...
+                            #include <rccl.h>
+                            #endif
+                            #include <stdio.h>
+                            #ifdef NCCL_MAJOR
+                            #ifndef NCCL_VERSION_CODE
+                            #  define NCCL_VERSION_CODE \
+                            (NCCL_MAJOR * 1000 + NCCL_MINOR * 100 + NCCL_PATCH)
+                            #endif
+                            #else
+                            #  define NCCL_VERSION_CODE 0
```
Comment (emcastillo) [inline install/build.py:424]: with rccl we will have the 0?

Comment (leofang) [reply install/build.py:424]: No, because `NCCL_VERSION_CODE`, `NCCL_MAJOR`, etc are all defined in `rccl.h`.

Comment (emcastillo) [reply install/build.py:424]: LOL they are defined as `NCCL` instead of `RCCL`?

Comment (leofang) [reply install/build.py:424]: That's right....

**Review by emcastillo — APPROVED** [review]

**Pull request merged**